spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/QueryPushdownIntegrationTestBase.java

```diff
@@ -432,4 +405,340 @@ public void testAggregateExpressions() {
+     |   3|   2|   1|[[[[2:str1, 3:str...|
+     +----+----+----+--------------------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      assertThat(r.size()).isEqualTo(4);
+    }
+    result =
+        df_to_join
+            .join(df, df.col("num1").equalTo(df_to_join.col("num1")), "leftsemi")
+            .collectAsList();
+    /*
+     +----+----+----+--------------------+
+     |num1|num2|num3|             strings|
+     +----+----+----+--------------------+
+     |   3|   4|   3|[[[[4:str4, 3:str...|
+     |   4|   1|   3|[[[[1:str1, 4:str...|
+     +----+----+----+--------------------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      assertThat(r.size()).isEqualTo(4);
+    }
+
+    result =
+        df.join(df_to_join, df.col("num1").equalTo(df_to_join.col("num1")), "leftanti")
+            .collectAsList();
+    /*
+     +----+----+----+-------+
+     |num1|num2|num3|strings|
+     +----+----+----+-------+
+     +----+----+----+-------+
+    */
+    assertThat(result.size()).isEqualTo(0);
+    result =
+        df_to_join
+            .join(df, df.col("num1").equalTo(df_to_join.col("num1")), "leftanti")
+            .collectAsList();
+    /*
+     +----+----+----+--------------------+
+     |num1|num2|num3|             strings|
+     +----+----+----+--------------------+
+     |   6|   5|   3|[[[[5:str5, 3:str...|
+     +----+----+----+--------------------+
+    */
+    assertThat(result.size()).isEqualTo(1);
+    assertThat(result.get(0).size()).isEqualTo(4);
+    assertThat(result.get(0).get(0)).isEqualTo(6);
+
+    result =
+        spark
+            .sql(
+                "SELECT "
+                    + "numStructDF.num1 AS a_num1, "
+                    + "numStructDF.num2 AS a_num2, "
+                    + "numStructDF.num3 AS a_num3, "
+                    + "numStructDF_to_join.num1 AS b_num1, "
+                    + "numStructDF_to_join.num2 AS b_num2, "
+                    + "numStructDF_to_join.num3 AS b_num3 "
+                    + "FROM numStructDF RIGHT JOIN numStructDF_to_join "
+                    + "ON numStructDF.num1 = numStructDF_to_join.num2 "
+                    + "WHERE numStructDF_to_join.num2 > 2 ")
+            .collectAsList();
+    /*
+     +------+------+------+------+------+------+
+     |a_num1|a_num2|a_num3|a_num2|b_num2|b_num3|
+     +------+------+------+------+------+------+
+     |  null|  null|  null|     6|     5|     3|
+     |     4|     3|     2|     3|     4|     1|
+     +------+------+------+------+------+------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      if (r.get(0) == null) {
+        assertThat(r.get(4)).isEqualTo(5);
+      } else {
+        assertThat(r.get(0)).isEqualTo(r.get(4));
+      }
+    }
+  }
+
+  /** Method to create a test table of schema NumStruct, in test dataset */
+  protected void writeTestDataToBigQuery() {
```

Comment:
Why don't we pass in the data to be written into this method? It would allow us to reuse this method

Reply:
Actually, I have done this in PR#744. But addressed here as well